core/src/main/java/com/linecorp/armeria/server/RoutingPredicate.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.linecorp.armeria.server.RoutingPredicate.ParsedComparingPredicate.parse;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.StreamSupport;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpParameters;
+
+/**
+ * A {@link Predicate} to evaluate whether a request can be accepted by a service method.
+ * Currently predicates for {@link HttpHeaders} and HTTP parameters are supported.
+ *
+ * @param <T> the type of the object to be tested
+ */
+public final class RoutingPredicate<T> {
```

Comment:
How does a user create a new instance of this class? I don't see any public static factory methods.

Reply:
OMG...Sorry about that. I thought I made some method public. But I think it would be better to make this class as private like before. Because we can add a predicate when we call `matches*` method.
https://github.com/line/armeria/pull/2102#discussion_r356055297